quil-rs/src/expression/simplification/by_hand.rs

```diff
@@ -160,671 +164,714 @@ fn mul_matches(left_ax: &Expression, right_ax: &Expression) -> bool {
+            // a / (-b) = (-a) / b, pick the shorter
+            (
+                ref left,
+                InfixOperator::Slash,
+                ref right @ Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    ref expression,
+                }),
+            ) => {
+                let original = div!(left.clone(), right.clone());
+                let neg_left = simplify_prefix(PrefixOperator::Minus, left, limit - 1);
+                let new = simplify_infix(&neg_left, InfixOperator::Slash, expression, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            // (-a) / b = a / (-b), pick the shorter
+            (
+                ref left @ Expression::Prefix(PrefixExpression {
+                    operator: PrefixOperator::Minus,
+                    ref expression,
+                }),
+                InfixOperator::Slash,
+                ref right,
+            ) => {
+                let original = div!(left.clone(), right.clone());
+                let neg_right = simplify_prefix(PrefixOperator::Minus, right, limit - 1);
+                let new = simplify_infix(expression, InfixOperator::Slash, &neg_right, limit - 1);
+                min_by_key(original, new, size)
+            }
+
+            //----------------------------------------------------------------
+            // Third: Affine relationships
+            //----------------------------------------------------------------
+
+            // (a1 * x + b1) + (a2 * x + b2) = (a1 + a2) * x + (b1 + b2)
+            //
+            // Apologies for this one; I couldn't get the compiler to let me match two levels deep in a
+            // recursive data type, and `if let` in a match guard isn't stabilized.
+            (
+                Expression::Infix(InfixExpression {
+                    left: ref left_ax,
+                    operator: InfixOperator::Plus,
+                    right: ref left_b,
+                }),
+                InfixOperator::Plus,
+                Expression::Infix(InfixExpression {
+                    left: ref right_ax,
+                    operator: InfixOperator::Plus,
+                    right: ref right_b,
+                }),
+            ) if mul_matches(left_ax, right_ax) => {
+                let &Expression::Infix(InfixExpression {
+                    left: ref ll,
+                    operator: InfixOperator::Star,
+                    right: ref lr,
+                }) = &**left_ax
+                else {
+                    unreachable!("This is handled by mul_matches")
+                };
+                let &Expression::Infix(InfixExpression {
+                    left: ref rl,
+                    operator: InfixOperator::Star,
+                    right: ref rr,
+                }) = &**right_ax
+                else {
+                    unreachable!("This is handled by mul_matches")
+                };
```

Comment:
Would it make more sense to have `mul_matches` such that you could write

```rust
if let Some((ll, lr, rl, rr)) = mul_matches(left_ax, right_ax) {
    // ...
}
```

Reply:
Oh, but `if let` in guard position in a `match` isn’t available on stable.

Reply:
Oh, I missed that that was in a guard position.